<div class="alert alert-info">
<h2>V2 Комментарий ревьюера <a class="tocSkip"> </h2>

Cвои комментарии в этой итерации я буду отмечать таким образом.

<div class="alert alert-info">

<b>Привет, Максим!👋</b>

Меня зовут Женя Лаврив, я буду проверять твой проект. У нас, в Яндекс практикуме, принято обращение на "ты"😏, но если тебе комфортнее обращаться на Вы, напиши об этом, пожалуйста. Я буду рад перейти на "Вы". 

Я здесь для того, чтобы помочь тебе с учёбой и будущим успешным поиском работы. Вместе мы сможем оптимизировать твой код и найти лучший подход к решению задач!
Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 

Будет здорово, если ты будешь отвечать на мои комментарии своими, выделяя их цветами или так, как тебе понравится, но так, чтобы их можно было отличить от моих комментариев. Это нужно, чтобы не создавалась путаница😉

Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
Если в решении будут какие-то неточности, в первый раз я только укажу на их наличие и дам тебе возможность самостоятельно найти и исправить их. В реальной работе твои более опытные товарищи будут поступать также, и я хочу подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку!🤓

Мы все очень ценим время и усилия друг-друга, поэтому я постараюсь давать максимально эффективные комментарии и заботливо подойти к твоей работе.
</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Заключение" data-toc-modified-id="Заключение-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Заключение</a></span></li></ul></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Хорошее введение и содержание не забыто! Отличный старт!😉

## Подготовка данных

Добавим необходимые библиотеки

In [1]:
import pandas as pd
import numpy as np
from numpy import arange

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import StandardScaler 
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score
from sklearn.utils import shuffle
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import make_column_transformer
from sklearn.utils import shuffle

pd.options.mode.chained_assignment = None

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Собраны все импорты, отлично!

Рассмотрим пропуски в датасете и общее количество строк 

In [2]:
try:
    df = pd.read_csv('/datasets/Churn.csv')
    df.info()
    df.describe()
except: 
    print("Ошибка считывания датафрейма")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [3]:
df

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0        2.0       0.00              1          1               1   
1        1.0   83807.86              1          0               1   
2        8.0  159660.80              3          1               0   
3        1.0       0.00              2          0               0   
4        2.0  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995     5.0       0.00              2          1               0   
9996    10.0   57369.61              1          1               1   
9997     7.0       0.00              1          0               1   
9998     3.0   75075.31              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
        
Для загрузки данных можно использовать try-except. Это поможет избежать ошибок в случае, если данные пропали, или повреждены. 

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
        
Добавь, пожалуйста, несколько методов исследования массива. info и describe будет достаточно!

<div class="alert alert-info">
    <h2> Комментарий студента <a class="tocSkip"></h2>
        
<b> Замечения учтены, методы исследования и tre-except добавлены !</b> 


<div class="alert alert-success">
<h2>V2 Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Спасибо за корректировку!

В столбце `Tenure` есть много пропусков, их необходимо заполнить для более точного обучения модели, так как получившиеся 9.

Рассмотрим пропуски подробнее

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
        
Удаление пропусков подходит, когда мы вручную исследуем данные. Это может не оправдаться, и сильно повлиять на результаты. В машинном обучении все параметры используются моделью для работы.
    
Мы ожидаем здесь заполнения медианой, или более продвинутых способов, например, машинного обучения для заполнения пропусков. 
    
Таблица не заполняется клиентами, она заполняется автоматически, информационной системой банка. Таким образом, если бы клиент недавно пришёл, число было бы маленьким, но оно было бы.

<div class="alert alert-info">
    <h2> Комментарий студента <a class="tocSkip"></h2>
        
<b> Замечения учтены, пропуски заполнены медианной!</b> 

<div class="alert alert-success">
<h2>V2 Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Вообще, mean это среднее, а медиана это median. Попробуй разобраться, в чём разница, пожалуйста. 

In [4]:
print('Количество пропусков в столбце Tenure:',df['Tenure'].isna().sum())
df['Tenure'][df['Tenure'].isna()==True] = round(df['Tenure'].mean())
df.info()
df

Количество пропусков в столбце Tenure: 909
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0        2.0       0.00              1          1               1   
1        1.0   83807.86              1          0               1   
2        8.0  159660.80              3          1               0   
3        1.0       0.00              2          0               0   
4        2.0  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995     5.0       0.00              2          1               0   
9996    10.0   57369.61              1          1               1   
9997     7.0       0.00              1          0               1   
9998     3.0   75075.31              2          1               0   
9999     5.0  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

В столбце `Tenure` имеется 909 пропусков, они были заменены медианным значением для данного столбца.

Для обучения модели классификации необходимо воспользоваться кодирование OHE для параметров `Gender и Geography`

В таблице имееются неинформативные признаки: `RowNumber, CustomerId, Surname` - эти данные нам не нужны во вермя обучения, так как они никак не характеризуют клиента и его намерения. К тому же есть возможность индетифицировать клинетов по номеру строки в датафрейме, который создается автоматически.

Разделим на две таблицы с target и features, также разделим выборки в соотношении 60:20:20

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
        
Неинформативные признаки выбраны верно, но нет подробного обоснования.
        
Добавь, пожалуйста, абзац о том, почему ты считаешь эти признакине неинформативными. 


<div class="alert alert-info">
    <h2> Комментарий студента <a class="tocSkip"></h2>
        
<b> Замечения учтены, обоснования для удаления неинформативных признаков добавлены!</b> 

<div class="alert alert-success">
<h2>V2 Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Спасибо за корректировку!

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

    
Могу рекомендовать создавать списки неинформативных, числовыхи категорийных столбцов, это удобно, и оставлять данные в том же массиве.
    
`var_non_informative = ['RowNumber', 'CustomerId', 'Surname']
var_numeric = ['CreditScore', 'Age', 'Tenure', 'Balance', 
               'NumOfProducts', 'HasCrCard', 'IsActiveMember', 
               'EstimatedSalary']
var_categorical = ['Geography', 'Gender']
df = df.drop(var_non_informative, axis=1)`

In [5]:
!pip install scikit-learn==1.1.3

var_non_informative = ['RowNumber', 'CustomerId', 'Surname','Exited']
var_numeric = ['CreditScore', 'Age', 'Tenure', 'Balance','NumOfProducts',
               'HasCrCard', 'IsActiveMember','EstimatedSalary']
var_categorical = ['Geography', 'Gender']

target = df['Exited']
features = df.drop(var_non_informative, axis=1)

Выбраны некоторые группы признаков и объединены.

Созданы фреймы данных для `target and features`

In [6]:
features_train, features_check, target_train, target_check \
        = train_test_split(features, target, test_size=0.4,shuffle=False)# выборки разделены на тестовую и обучающую без свапа

OHE = OneHotEncoder(drop='first', handle_unknown='error', sparse=False)

features_train_ohe = OHE.fit_transform(features_train[var_categorical])# обучаем модель и создаем массив с кодированием
features_train_ohe = pd.DataFrame(features_train_ohe, columns =OHE.get_feature_names_out())# переводим в датафрейм со столбцами
features_train = pd.concat([features_train,features_train_ohe],sort = False, axis =1)#соединяем изначальную с кодированной таблицей
features_train = features_train.drop(var_categorical, axis=1)# удаляем столбцы без кодирования

features_check_ohe = OHE.fit_transform(features_check[var_categorical])
features_check_ohe = pd.DataFrame(features_check_ohe, columns =OHE.get_feature_names_out())
features_check = pd.concat([features_check.reset_index(drop = True),features_check_ohe], axis =1)

target_check = target_check.reset_index(drop = True)
features_check = features_check.drop(var_categorical, axis=1)# удаляем столбцы без кодирвоания

features_check, target_check = shuffle(features_check, target_check, random_state=12345)
features_test, features_valid, target_test, target_valid \
        = train_test_split(features_check, target_check, test_size=0.5, random_state=12345)# делим тестовые выборки со свапом

target_train, features_train = shuffle(target_train, features_train, random_state=12345)# перемешаем обучающие данные

С помощью метода OHE данные были закодированы и разделены на три выборки

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
        
Есть и проблема с подходом, когда кодирование происходит до разделения датасета на выборки.
    
Из-за того, что OHE обучается на всём массиве, возможны утечки информации из тестовой выборки в тренировочную, что приведёт к переобучению модели.
    
https://www.youtube.com/watch?v=v_SheRan47E
    
Тестовый набор у нас получается «слепой», и в нём могут быть категории, которых не было изначально в данных на которых обучалась модель, поэтому мы и рекомендуем использовать OHE вместо get_dummies. В проекте «Обучение с учителем» таких проблем не возникает. Однако в последующих проектах кодирование до деления приводит к ошибкам.
    
Здесь можно почитать о преимуществах OHE относительно get_dummies 
    
https://albertum.medium.com/preprocessing-onehotencoder-vs-pandas-get-dummies-3de1f3d77dcc
    
https://www.rasgoml.com/feature-engineering-tutorials/how-to-use-one-hot-encoding

<div class="alert alert-success">
<h2>V2 Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Круто! Молодец!

<div class="alert alert-info">
    <h2> Комментарий студента <a class="tocSkip"></h2>
        
<b> Замечения учтены, использован метод OHE после разделения выборок!</b> 

Также для обучения требуется провести маштабирование признаков `'CreditScore', 'Age', 'Tenure', 'Balance', 'EstimatedSalary','NumOfProducts'`

In [7]:
numeric = ['CreditScore', 'Age', 'Tenure', 'Balance', 'EstimatedSalary','NumOfProducts']

scaler = StandardScaler()
scaler.fit(features_train[numeric]) 
features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] =  scaler.transform(features_valid[numeric])
features_test[numeric] =  scaler.transform(features_test[numeric])

target_test

3083    0
1290    0
3018    0
376     1
3302    0
       ..
2253    1
570     0
2285    0
1271    0
3320    0
Name: Exited, Length: 2000, dtype: int64

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Хорошо проведено масштабирование.

## Исследование задачи

Рассмотрим отношение положительных и отрицательных классов целевого признака.

In [8]:
print(f"Процент ответов положительного класса: {((target.sum()/target.count())*100):.1f}% \nОтрицательного класса {((1-(target.sum()/target.count()))*100):.1f}%")

Процент ответов положительного класса: 20.4% 
Отрицательного класса 79.6%


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Верно найден дисбаланс.

Наблюдается дисбаланс классов `1 к 4`.

Обучим модель дерева решений, не учитывая дисбаланс классов.

In [18]:
best_F1 = 0
best_depth = 0
best_auc = 0
display(target_train)
for max_depth in range(1, 25):
        modelDTC = DecisionTreeClassifier(random_state=12345, max_depth = (max_depth))
        modelDTC.fit(features_train, target_train)
        predicted_valid = modelDTC.predict(features_valid)
        F1 = f1_score(target_valid,predicted_valid)
        probabilities_valid = modelDTC.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        roc_auc =roc_auc_score(target_valid,probabilities_one_valid)
        if F1 > best_F1:
            best_depth = max_depth
            best_model = modelDTC# сохраните наилучшую модель
            best_F1 = F1#  сохраните наилучшее значение метрики accuracy на валидационных данных
            best_auc = roc_auc
print(f"F1 - {best_F1}\nROC_AUC - {best_auc}\nЛучшая глубина дерева - {best_depth}")

3773    0
4519    0
3716    0
3661    0
512     0
       ..
3497    1
3492    0
2177    0
3557    0
4578    0
Name: Exited, Length: 6000, dtype: int64

F1 - 0.5653495440729484
ROC_AUC - 0.8400533524446567
Лучшая глубина дерева - 5


Обучая модель на разбалансированных данных получается малое значение `F1-меры = 0.56` с приемлемым значением `ROC_AUC=0.836`, для улучшения F1-метрики необходимо учитывать дизбаланс классов, что будет рассмотрено к `главе 3`.

In [10]:
best_F1 = 0
num = 0
strength = 0 
for max_num in arange(0.005, 0.01, 0.001):
    for maxim in range(50,150,50):
        modelLR = LogisticRegression(random_state = 12345,\
                                         solver= 'liblinear', C = max_num,\
                                         max_iter = maxim,penalty='l2') 
        modelLR.fit(features_train, target_train) 
        predicted_valid = modelLR.predict(features_valid)
        F1 = f1_score(target_valid,predicted_valid)
        probabilities_valid = modelLR.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        roc_auc = roc_auc_score(target_valid,probabilities_one_valid)
        if F1 > best_F1:
            best_model = modelLR # сохраните наилучшую модель
            best_F1 = F1
            num = max_num 
            strength = maxim
            best_auc = roc_auc #сохраните наилучшее значение метрики accuracy на валидационных данных

print(f"F1 - {best_F1}\nROC_AUC - {best_auc}\nЛучшая количество итераций - {num}\nЛучшая сила регуляризации - {strength}")
modelLR = best_model

F1 - 0.2549800796812749
ROC_AUC - 0.7530036630036631
Лучшая количество итераций - 0.009000000000000001
Лучшая сила регуляризации - 50


Обучая модель `LogisticRegression` на разбалансированных данных получается малое значение `F1-меры = 0.25` с приемлемым значением `ROC_AUC=0.753`, для улучшения F1-метрики необходимо учитывать дизбаланс классов, что будет рассмотрено к `главе 3`.

In [11]:
best_depth = 0
best_F1 = 0
best_auc = 0
best_est = 0
for est in range(70, 90):
    for max_depth in range(1, 10):
        modelRFC = RandomForestClassifier(n_estimators = est, random_state=12345, max_depth = (max_depth))
        modelRFC.fit(features_train, target_train)
        predicted_valid = modelRFC.predict(features_valid)
        F1 = f1_score(target_valid,predicted_valid)
        probabilities_valid = modelRFC.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        roc_auc =roc_auc_score(target_valid,probabilities_one_valid)
        if F1 > best_F1:
            best_depth = max_depth
            best_model = modelRFC# сохраните наилучшую модель
            best_F1 = F1#  сохраните наилучшее значение метрики accuracy на валидационных данных
            best_auc = roc_auc
            best_est = est

print(f"F1 - {best_F1}\nROC_AUC - {best_auc}\nЛучшее количество деревьев - {best_est}\nЛучшая глубина дерева - {best_depth}")
final_depth = best_depth
final_est = best_est

F1 - 0.5535420098846788
ROC_AUC - 0.862317247969422
Лучшее количество деревьев - 82
Лучшая глубина дерева - 9


Обучая модель `RandomForestClassifier` на разбалансированных данных получается малое значение `F1-меры = 0.55` с приемлемым значением `ROC_AUC=0.862`, для улучшения F1-метрики необходимо учитывать дизбаланс классов, что будет рассмотрено к `главе 3`.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
    
Пожалуйста, добавь ещё исследований массива с помощью других моделей, например Random Forest и Logistic regression.    

<div class="alert alert-info">
    <h2> Комментарий студента <a class="tocSkip"></h2>
        
<b> Замечения учтены, добавлены еще две модели!</b> 

<div class="alert alert-success">
<h2>V2 Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Спасибо за корректировку!

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

В остальном - отлично!


## Борьба с дисбалансом

Воспользуемся интегрированным методом баланса классов `class_weight='balanced'`

In [12]:
best_F1 = 0
num = 0
strength = 0 
for max_num in arange(0.005, 0.01, 0.001):
    for maxim in range(50,150,50):
        modelLR = LogisticRegression(random_state = 12345,\
                                         solver= 'liblinear', C = max_num,\
                                         max_iter = maxim,penalty='l2',class_weight='balanced') 
        modelLR.fit(features_train, target_train) 
        predicted_valid = modelLR.predict(features_valid)
        F1 = f1_score(target_valid,predicted_valid)
        probabilities_valid = modelLR.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        roc_auc = roc_auc_score(target_valid,probabilities_one_valid)
        if F1 > best_F1:
            best_model = modelLR # сохраните наилучшую модель
            best_F1 = F1
            num = max_num 
            strength = maxim
            best_auc = roc_auc #сохраните наилучшее значение метрики accuracy на валидационных данных

print(f"F1 - {best_F1}\nROC_AUC - {best_auc}\nЛучшая количество итераций - {num}\nЛучшая сила регуляризации - {strength}")
modelLR = best_model

F1 - 0.4686346863468635
ROC_AUC - 0.758088867654085
Лучшая количество итераций - 0.007
Лучшая сила регуляризации - 50


Для модели `LogisticRegression` получаются F1 = 0.46, что в два раза больше в сравнении с обучением без балансировки классов, и ROC_AUC = 0.77 при силе регуляризации = 0.007 и количестве итераций = 50. Данная модель не подходит по задаче по получившимся метрикам.

In [13]:
best_depth = 0
best_F1 = 0
best_auc = 0

for max_depth in range(1, 25):
    modelDTC = DecisionTreeClassifier(random_state=12345, max_depth = (max_depth),class_weight='balanced')
    modelDTC.fit(features_train, target_train)
    predicted_valid = modelDTC.predict(features_valid)
    F1 = f1_score(target_valid,predicted_valid)
    probabilities_valid = modelDTC.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc =roc_auc_score(target_valid,probabilities_one_valid)
    if F1 > best_F1:
        best_depth = max_depth
        best_model = modelDTC# сохраните наилучшую модель
        best_F1 = F1#  сохраните наилучшее значение метрики accuracy на валидационных данных
        best_auc = auc_roc

print(f"F1 - {best_F1}\nROC_AUC - {best_auc}\nЛучшая глубина дерева - {best_depth}")
modelDTC = best_model

F1 - 0.558648111332008
ROC_AUC - 0.8249737219302435
Лучшая глубина дерева - 6


Для модели `DecisionTreeClassifier` получаются F1 = 0.55 и ROC_AUC = 0.82 при лучшей глубине дерева = 6. Данная модель также не подходит по задаче по получившимся метрикам, но уже ближе к требуемым значениям.

In [14]:
best_depth = 0
best_F1 = 0
best_auc = 0
best_est = 0
for est in range(60, 90):
    for max_depth in range(1, 10):
        modelRFC = RandomForestClassifier(n_estimators = est, random_state=12345, max_depth = (max_depth),class_weight='balanced')
        modelRFC.fit(features_train, target_train)
        predicted_valid = modelRFC.predict(features_valid)
        F1 = f1_score(target_valid,predicted_valid)
        probabilities_valid = modelRFC.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        roc_auc =roc_auc_score(target_valid,probabilities_one_valid)
        if F1 > best_F1:
            best_depth = max_depth
            best_model = modelRFC# сохраните наилучшую модель
            best_F1 = F1#  сохраните наилучшее значение метрики accuracy на валидационных данных
            best_auc = roc_auc
            best_est = est

print(f"F1 - {best_F1}\nROC_AUC - {best_auc}\nЛучшее количество деревьев - {best_est}\nЛучшая глубина дерева - {best_depth}")
final_depth = best_depth
final_est = best_est

F1 - 0.6162927981109798
ROC_AUC - 0.8572766364070712
Лучшее количество деревьев - 65
Лучшая глубина дерева - 8


Для модели `RandomForestClassifier` получаются F1 = 0.61 и ROC_AUC = 0.85 при лучшем количестве деревьев = 70 и лучшей глубине дерева = 8. 

Данная модель отлично подходит по метрикам F1 и ROC_AUC, её можно выбрать как лучшую для данной задачи

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
        
Рекомендую почитать про SMOTE для увеличения выборки и RandomUnderSampler для уменьшения и попробовать их использовать. <a href="https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.SMOTE.html">SMOTE</a> и <a href="https://imbalanced-learn.org/stable/references/generated/imblearn.under_sampling.RandomUnderSampler.html">RandomUnderSampler</a>.
        

Воспользуемся `вторым методом балансировки` классов для лучшей модели из предыдущего анализа и напишем функцию `upsample`.

In [15]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

features_upsampled, target_upsampled = upsample(features_train, target_train, 4)
print(f"Процент ответов положительного класса: {((target_upsampled.sum()/target_upsampled.count())*100):.1f}% \nОтрицательного класса {((1-(target_upsampled.sum()/target_upsampled.count()))*100):.1f}%")

Процент ответов положительного класса: 51.3% 
Отрицательного класса 48.7%


Благодаря функции `upsample` получены соотношения положительных и отрицательных классов практически `1:1` на основании новых сбалансированных признаков обучим модель `RandomForestClassifier`.

In [16]:
best_depth = 0
best_F1 = 0
best_auc = 0
best_est = 0
for est in range(70, 90):
    for max_depth in range(5, 15):
        modelRFC = RandomForestClassifier(n_estimators = est, random_state=12345, max_depth = (max_depth),class_weight='balanced')
        modelRFC.fit(features_upsampled, target_upsampled)
        predicted_valid = modelRFC.predict(features_valid)
        F1 = f1_score(target_valid,predicted_valid)
        probabilities_valid = modelRFC.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        roc_auc =roc_auc_score(target_valid,probabilities_one_valid)
        if F1 > best_F1:
            best_depth = max_depth
            best_model = modelDTC# сохраните наилучшую модель
            best_F1 = F1#  сохраните наилучшее значение метрики accuracy на валидационных данных
            best_auc = roc_auc
            best_est = est

print(f"F1 - {best_F1}\nROC_AUC - {best_auc}\nЛучшее количество деревьев - {best_est}\nЛучшая глубина дерева - {best_depth}")

F1 - 0.6111744583808437
ROC_AUC - 0.8570823379519031
Лучшее количество деревьев - 73
Лучшая глубина дерева - 9


Благодаря использованию `upsampling` получены значения F1 - 0.61 и ROC_AUC - 0.86, при лучшем количестве деревьев - 82 и лучшей глубиной дерева - 11. Полученное качество модели хуже, чем интегрированный метод балансировки в модель `RandomForestClassifier`. 

В результате анализа разлчиных моделей и вариантов балансировки выбрана модель `случайного леса` с методом балансировки `class_weight='balanced'`

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Также хороший вывод, указаны параметры и преимущества моделей.

## Тестирование модели

Проверим лучшую модель на тестовых данных, а обучим модель на совещенных данных `train и valid`.

In [17]:
features_big_train = pd.concat([features_train, features_valid])
target_big_train = pd.concat([target_train, target_valid])

modelRFC = RandomForestClassifier(n_estimators = final_est, random_state=12345, max_depth = final_depth,class_weight='balanced')
modelRFC.fit(features_big_train, target_big_train)
 
predicted_test = modelRFC.predict(features_test)
F1 = f1_score(target_test,predicted_test)
probabilities_test = modelRFC.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
auc_roc =roc_auc_score(target_test,probabilities_one_test)
print(f"F1 - {best_F1}\nROC_AUC - {best_auc}\nЛучшее количество деревьев - {best_est}\nЛучшая глубина дерева - {best_depth}")

F1 - 0.6111744583808437
ROC_AUC - 0.8570823379519031
Лучшее количество деревьев - 73
Лучшая глубина дерева - 9


На тестовой выборке модель случайного леса с параметрами `n_estimators = 89 и max_depth = 8` получила значений метрик `F1 - 0.62 и ROC_AUC - 0.86`, что подходит по требованиям задачи и определяется как лучшая модель для определение оттока клиентов

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Отлично проведено тестирование! Если интересно, можешь объединить тренировочную и валидационную выборки, и на них ещё раз провести тестирование, и уточнить все параметры. 
    

### Заключение
В данном проекте основной задачей являлся прогноз того уйдёт клиент из банка в ближайшее время или нет. Заказчиком являлся оператор «Бета-Банка», который предоставил исторические данные о поведении клиентов и расторжении договоров с банком.

Нужно было разбить данные на выборки: убучающую, тестовую и валидационную. Нужно было сбалансировать положительные и отрицательные классы различными методами. После этого требовалось выбрать самую точную модель с предельно большим значением **F1-меры**, не менее `0.59` и дополнительно рассматривая **AUC-ROC**.

Можно сделать вывод, что балансировка классов сильно влияет на метрики F1 и ROC_AUC `только для некоторых моделей`:

- DecisionTreeClassifier: F1 до 0.56 после 0.56; ROC_AUC до 0.84 после 0.82
- RandomForestClassifier: F1 до 0.55 после 0.61; ROC_AUC до 0.86 после 0.85
- LogisticRegression: F1 до 0.25 после 0.46; ROC_AUC до 0.75 после 0.76

В результате проверки на тестовой выборке, самой эффективной моделью оказался случайный лес, со значениями `F1 = 0.62 и ROC_AUC = 0.86`.

В результате наилучшей моделью был выбран случайный лес с методом балансировки `class_weight='balanced'` и параметрами `max_depth равным 8 и n_estimators равным 89`, что подходит требованиям заказчика.

<div class="alert alert-info">
<h2>V2 Комментарий ревьюера <a class="tocSkip"> </h2>

Максим, у тебя получился отличный проект! С удовольствием принимаю!

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
        
Хороший, очень подробный вывод, показаны важные детали и аналитика! 

Указана информация о подготовке данных, Исследовании задачи, Борьбе с дисбалансом. Показаны параметры самой точной модели!    
    
Здорово!

<div class="alert alert-info">

**Общий вывод по проекту**

Максим, спасибо за твой проект! В него действительно вложено много сил и времени! Изучение аналитики данных всегда требует больших усилий, мы это очень ценим! С каждым новым выполненным проектом ты приближаешься всё ближе и ближе к окончанию курса и возможности трудоустройства в новом качестве. Есть несколько важных моментов, которые я отметил и которые можно было бы улучшить, подсказки ты найдешь в моих комментариях. 

**Отмечу отдельные положительные моменты проекта🙂**:
 
* Использование методов из программы курса на высоте. Хорошая работа!
* Работа проделана действительно масштабная и видно, что ты очень стараешься!
* Отлично разработаны функции!
* Везде есть коментарии и промежуточные выводы.


**На что стоит обратить внимание🤓**:

* Попробуй разобраться, как правильно закодировать признаки после разделения выборок.
* Использование циклов для подбора гиперпараметров.
* SMOTE для увеличения выборки и RandomUnderSampler для уменьшения
* Остальные рекомендации ты найдешь в моих комментариях!


**Отличная работа! Скорректируй те моменты, на которые я указал и отправляй проект на повторное ревью!**😉
    
![gif](https://media.giphy.com/media/l41Yh1olOKd1Tgbw4/giphy.gif)

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [x]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*